components/lib/dataprotect/src/main/java/mozilla/components/lib/dataprotect/SecureAbove22Preferences.kt

```diff
@@ -0,0 +1,168 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.lib.dataprotect
+
+import android.annotation.TargetApi
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+import android.content.SharedPreferences
+import android.os.Build
+import android.os.Build.VERSION_CODES.M
+import android.util.Base64
+import androidx.annotation.VisibleForTesting
+import mozilla.components.support.base.log.logger.Logger
+import java.nio.charset.StandardCharsets
+import java.security.GeneralSecurityException
+import java.security.KeyStore
+
+/**
+ * A wrapper around [SharedPreferences] which encrypts contents on supported API versions (23+).
+ * Otherwise, this simply delegates to [SharedPreferences].
+ *
+ * In rare circumstances (such as APK signing key rotation) a master key which protects this storage may be lost,
+ * in which case previously stored values will be lost as well. Applications are encouraged to instrument such events.
+ *
+ * @param context A [Context], used for accessing [SharedPreferences].
+ */
+class SecureAbove22Preferences(context: Context) {
+    companion object {
+        const val KEY_PREFERENCES_POST_M = "key_preferences_post_m"
+        const val KEY_PREFERENCES_PRE_M = "key_preferences_pre_m"
+
+        private const val BASE_64_FLAGS = Base64.URL_SAFE or Base64.NO_PADDING
+    }
+
+    private val keystore: Keystore? = if (Build.VERSION.SDK_INT < M) {
+        null
+    } else {
+        Keystore(context.packageName)
+    }
```

Comment:
nit: There are a lot of `if build version < m else` blocks in this class. Those make it quite hard to follow and mixing that with nullability (`keystore!!`) makes it easy to introduce errors. The support libraries run into this case a lot where they have to support different implementations for different API levels. They solve this nicely by using delegation. In this case they'd delegate internally to something like SecurePreferences21 (for 21+) and SecurePreferences23 (for 23+). Here this would make this code much easier to read and avoid some nullability issues.

An example:
* NotificationCompat: https://chromium.googlesource.com/android_tools/+/refs/heads/master/sdk/sources/android-25/android/support/v4/app/NotificationCompat.java#939
* NotificationCompatApi21: https://chromium.googlesource.com/android_tools/+/refs/heads/master/sdk/sources/android-25/android/support/v4/app/NotificationCompatApi21.java
* ...

Reply:
Re-wrote the class following this pattern, and it's _much_ better now. Thanks!